PROBLEMS.md

```diff
@@ -0,0 +1,23 @@
+While IdentityCache allows to massively scale access to ActiveRecord objects, it's important to note what problems might arise with it over time and what are things that it's not solving in the current design.
+
+## Large blobs
+
+For [god objects](https://en.wikipedia.org/wiki/God_object) (like the Shop model at Shopify), it’s typical to have many associations. It’s for developers appealing to embed many of those associations into the god model’s IDC blob with `cache_has_many :images, embed: true`. As a side effect, over time the god model’s IDC blob may grow over the top. It gets even worse when some of the embedded records get larger, and the total IDC blob gets into as much as a megabyte in size. This puts pressure on both the application and memcached, because the blob needs to be serialized and unserialized each time and transferred over the network.
+
+## Overfetching
+
+For those fat (and possibley god) objects, every `Model.fetch` would result in fetching all of the blob over the network and unserializing all embedded records, even if the client has only accessed one field. That's a lot of useless CPU cycles and IO operations that happen on each access.
+
+## Hot keys
+
+This is not necessarily the IDC problem, but it is a problem with memcache and the way how IDC uses it. To scale out memcache, it’s typical to arrange memcached instances in rings. Once a memcache operation comes in, [consistent hashing](https://github.com/facebook/mcrouter/wiki/Pools#hash-functions) is used to route the operation to one of the rings. So a key that maps for the same IDC entry will always end in the same memcached instance.
+
+If that key associated with a record becomes busy and hot, that memcached instance will become saturated on the network and the number of bytes that it can send out. In cloud environments, it’s a limit that is easy to hit.
+
+## Thundering herd
```

Comment:
This one seems like it is more clearly just an issue that needs fixing.  We even have a pull request with a possible fix (https://github.com/Shopify/identity_cache/pull/373)